File: crates/graphforge-bindings-py/tests/branches.py
"""Native Branch isolation, durable replay and Arrow parity."""

from pathlib import Path
import tempfile
import unittest
from uuid import uuid4

import graphforge


def identity():
    return str(uuid4())


class BranchTests(unittest.TestCase):
    def test_independent_branch_restore_replay_and_reopen(self):
        with tempfile.TemporaryDirectory() as directory:
            root = str(Path(directory) / "project")
            graph = graphforge.GraphForge(root)
            graph.execute("CREATE (:Character {name: 'Original'})")

            def generation():
                return graph.research_project_summary()["identity"]["generation_uuid"]

            branch, base = identity(), identity()
            create = {
                "operation_uuid": identity(),
                "expected_generation_uuid": generation(),
                "branch_uuid": branch,
                "version_uuid": base,
                "source": {
                    "kind": "current",
                    "origin_version_uuid": identity(),
                    "context_uuid": identity(),
                },
                "creator_uuid": identity(),
                "created_at": 1,
                "label": "Story",
            }
            receipt = graph.create_research_branch(create)
            self.assertEqual(graph.create_research_branch(create), receipt)
            selection = graph.research_branch_selection(branch)
            self.assertEqual(selection.num_rows, 1)
            edit = {
                "operation_uuid": identity(),
                "expected_generation_uuid": generation(),
                "branch_uuid": branch,
                "version_uuid": identity(),
                "created_at": 2,
                "query": "MATCH (n:Character) SET n.name = 'Local'",
            }
            graph.execute_research_branch(edit)
            self.assertEqual(
                graph.query_research_branch(branch, "MATCH (n) RETURN n.name AS name")["name"][
                    0
                ].as_py(),
                "Local",
            )
            self.assertEqual(
                graph.execute("MATCH (n) RETURN n.name AS name")["name"][0].as_py(), "Original"
            )
            self.assertTrue(selection.equals(graph.research_branch_selection(branch)))
            self.assertIn("local", graph.research_branch_fields(branch)["status"].to_pylist())
            token = graphforge.CancellationToken()
            token.cancel()
            restore = {
                "operation_uuid": identity(),
                "expected_generation_uuid": generation(),
                "branch_uuid": branch,
                "source_version_uuid": base,
                "version_uuid": identity(),
                "created_at": 3,
            }
            with self.assertRaises(Exception) as error:
                graph.restore_research_branch(restore, cancellation=token)
            self.assertEqual(error.exception.code, "GF_CANCELLED")
            restored = graph.restore_research_branch(restore)
            graph.close()
            graph = graphforge.GraphForge(root)
            self.assertEqual(graph.restore_research_branch(restore), restored)
            self.assertEqual(graph.research_branch(branch)["version_uuid"], restore["version_uuid"])
            self.assertEqual(
                graph.query_research_branch(branch, "MATCH (n) RETURN n.name AS name")["name"][
                    0
                ].as_py(),
                "Original",
            )
            self.assertEqual(graph.research_branch_references(branch).num_rows, 0)
            graph.close()


def main():
    suite = unittest.defaultTestLoader.loadTestsFromTestCase(BranchTests)
    result = unittest.TextTestRunner().run(suite)
    if not result.wasSuccessful():
        raise SystemExit(1)


if __name__ == "__main__":
    main()

File: crates/graphforge-cli/tests/branches.rs
//! Real CLI Branch publication and immutable Arrow reads across process reopen.
use graphforge_api::*;
use std::{io::Cursor, path::Path, process::Command};
use uuid::Uuid;
fn run(root: &Path, args: &[&str]) -> std::process::Output {
    Command::new(env!("CARGO_BIN_EXE_gf"))
        .arg("--project")
        .arg(root)
        .args(["research", "branch"])
        .args(args)
        .output()
        .unwrap()
}
fn success(output: std::process::Output) -> Vec<u8> {
    assert!(
        output.status.success(),
        "{}",
        String::from_utf8_lossy(&output.stderr)
    );
    output.stdout
}
#[test]
fn cli_branch_edits_restore_without_changing_parent() {
    let directory = tempfile::tempdir().unwrap();
    let root = directory.path().join("project");
    let graph = GraphForge::new(root.to_str()).unwrap();
    graph
        .execute("CREATE (:Character {name:'Original'})")
        .unwrap();
    let branch = Uuid::now_v7();
    let base = Uuid::now_v7();
    let generation = graph
        .research_project_summary()
        .unwrap()
        .identity
        .generation_uuid;
    drop(graph);
    let path = directory.path().join("request.json");
    let file = path.to_str().unwrap();
    let create = serde_json::json!({"operation_uuid":Uuid::now_v7(), "expected_generation_uuid":generation, "branch_uuid":branch, "version_uuid":base, "source":{"kind":"current","origin_version_uuid":Uuid::now_v7(),"context_uuid":Uuid::now_v7()},"creator_uuid":Uuid::now_v7(),"created_at":1,"label":"Story"});
    std::fs::write(&path, serde_json::to_vec(&create).unwrap()).unwrap();
    let receipt = success(run(&root, &["create", "--file", file]));
    assert_eq!(success(run(&root, &["create", "--file", file])), receipt);
    let id = branch.to_string();
    let selected = success(run(&root, &["selection", "--branch-uuid", &id]));
    let count: usize = arrow::ipc::reader::StreamReader::try_new(Cursor::new(&selected), None)
        .unwrap()
        .map(|b| b.unwrap().num_rows())
        .sum();
    assert_eq!(count, 1);
    let receipt: serde_json::Value = serde_json::from_slice(&receipt).unwrap();
    let edit = serde_json::json!({"operation_uuid":Uuid::now_v7(),"expected_generation_uuid":receipt["generation_uuid"],"branch_uuid":branch,"version_uuid":Uuid::now_v7(),"created_at":2,"query":"CREATE (:Character {name:'Local'})"});
    std::fs::write(&path, serde_json::to_vec(&edit).unwrap()).unwrap();
    let edited: serde_json::Value =
        serde_json::from_slice(&success(run(&root, &["execute", "--file", file]))).unwrap();
    let result = success(run(
        &root,
        &[
            "query",
            "--branch-uuid",
            &id,
            "--query",
            "MATCH (n) RETURN n",
        ],
    ));
    assert_eq!(
        arrow::ipc::reader::StreamReader::try_new(Cursor::new(result), None)
            .unwrap()
            .map(|b| b.unwrap().num_rows())
            .sum::<usize>(),
        2
    );
    assert_eq!(
        success(run(&root, &["selection", "--branch-uuid", &id])),
        selected
    );
    let restore = serde_json::json!({"operation_uuid":Uuid::now_v7(),"expected_generation_uuid":edited["generation_uuid"],"branch_uuid":branch,"source_version_uuid":base,"version_uuid":Uuid::now_v7(),"created_at":3});
    std::fs::write(&path, serde_json::to_vec(&restore).unwrap()).unwrap();
    success(run(&root, &["restore", "--file", file]));
    let graph = GraphForge::new(root.to_str()).unwrap();
    assert_eq!(
        graph
            .execute("MATCH (n) RETURN n")
            .unwrap()
            .batches
            .iter()
            .map(|b| b.num_rows())
            .sum::<usize>(),
        1
    );
    assert_eq!(
        graph
            .open_research_branch(branch)
            .unwrap()
            .graph()
            .execute("MATCH (n) RETURN n")
            .unwrap()
            .batches
            .iter()
            .map(|b| b.num_rows())
            .sum::<usize>(),
        1
    );
}
